### ISSUE TITLE
language: specify if type literals are valid const map keys

### ISSUE THREAD

**Issue by rakudrama** [dart-lang/sdk#21553]
The spec as I read is does not tell us whether or not the following program is legal. It depends on whether the implementation of Type overrides `operator==()`. The VM and dart2js differ in this regard which is a problem for portability as programs developed on the VM don't work on dart2js (#17207)

It would be helpful to have a ruling on whether this specific combination of features is supposed to work and then have it implemented uniformly.

```dart
class A {}

const table = const {
    A: 'Apple',
};

main() {
    print(table[A]);
}
```

Wrapping the type to avoid operator== as follows leads to surprising behaviour.

```dart
class A {}
class B {}

class Key {
    final Type type;
    const Key(this.type);
}

const table = const {
    const Key(A): 'Apple',
    const Key(B): 'Banana',
};

main() {
    print(table[const Key(A)]);  // Apple
    print(table[new Key(B)]);    // null
}
```

Related issues: #17123 #17207

**Comment by hterkelsen** [comment]
I think that Type should be allowed as const Map keys, and there should be an exception made for them in the spec as there is with Symbol literals.

**Comment by lrhn** [comment]
If we prevent `Type` from being implementable (like `int`, `String`, etc.) it would probably make it easier to special-case it without also special-casing user implementations.

**Comment by leafpetersen** [comment]
Also note this use case from #30257 , which is abstracted from real code that I ran into internally.

```dart
class  A {}

void main() {
  var x = new A();
  switch (x.runtimeType) {
    case A: print("hello");
  }
}
```

cc @lrhn @eernstg @munificent @floitschG

**Comment by eernstg** [comment]
Note that we have abstained from introducing the "user can't extend/implement" constraint on `Type` on earlier occasions. The concrete reason was that the package reflectable needs to create "fake types" for `Type` values that our current run-time semantics does not support obtaining without reflection. This may indicate that reflectable is special (so we can have that constraint for "normal" user code anyway), or it may indicate that other "reflection-ish" software may need to create "fake types" as well. Or we may add some other way to create those fake types.

**Comment by matanlurey** [comment]
FWIW AngularDart is starting to remove all support for using `Type` at runtime to mean "lookup" something because of the code size effect of supporting "reflection-ish" scenarios. We'd strongly support whatever language option makes our code the smallest over whatever adds more features.

**Comment by lowagner** [comment]
I'm writing a flutter app with some tests, and ran into this issue with the switch case, using class inheritance...

i would like to do something like this:
```
class Parent {}
class Child1 extends Parent {
 int value;
}
class Child2 extends Parent {
 double data;
}
Parent x = Child1();
switch (x.runtimeType) {
 case Child1: doSomethingWith(x.value); break;
 case Child2: doSomethingElseWith(x.data); break;
}
```
however, the block does not recognize that x must be Child1 inside of that case, so it errors on accessing x.value.  Strangely enough, it works if i put extra logic in the block:
```
switch (x.runtimeType) {
 case Child1:
  if (x is Child1) doSomethingWith(x.value);
  break;
 ...
}
```
is there a way Dart can be smarter about inferring the types?  i understand there's an issue if we use labels in the switch case (or continues), but barring those, we should be able to infer types in these case blocks.

a related issue is that if we're in a block and we check something like this:
```
for (Parent p in [Child1(), Child2()]) {
 if (!(p is Child1)) continue;
 // do something with p as Child1:
 print(p.value); // ERROR, compiler doesn't see p as type Child1
}
```
in this block, after the continue, we should be able to infer the type of p as Child1, but no such luck...

**Comment by zoechi** [comment]
@lrhn wasn't overriding `runtimeType` removed in Dart 2?

**Comment by lrhn** [comment]
No, I don't think that was ever planned. We were hoping to remove `runtimeType` *completely*, and requiring you to use a new `Type.of` method/constructor instead, but our breaking-change vs importance budget wasn't high enough for that. We may still introduce `Type.of` at a later time, so there is an alternative to `runtimeType`, but unless we also allow subtype checks between `Type` objects, their usefulness really isn't that great.

My recommendation still stands: Never use `runtimeType`. :)

**Comment by lrhn** [comment]
To allow `Type` literals as const map keys (or switch case expressions), the specification needs to be changed to state that it allows objects created from type literals.

It's possible to implement `Type` both with and without overriding `==`, so we cannot let the implementation decide. It should be assumed, and currently is, that `Type` objects do override `==`, so you can't use `Type` objects.

In general, we need to specify that for all objects that can be constant expressions and where the actual implementation class is platform specific.

**Comment by eernstg** [comment]
Agreed; we have separate threads about how to determine equality of reified types, and the title of this issue is a question which was resolved in 0808d29b4e33c303bf40524407a6e331c7a439f8.
